cpp/python/include/ucxx/python/python_future_task.h

```diff
@@ -0,0 +1,213 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#pragma once
+
+#include <chrono>
+#include <future>
+#include <memory>
+#include <queue>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <iostream>
+
+#include <Python.h>
+
+#include <ucxx/log.h>
+#include <ucxx/python/future.h>
+
+namespace ucxx {
+
+namespace python {
+
+template <typename ReturnType, typename... TaskArgs>
+class PythonFutureTask : public std::enable_shared_from_this<PythonFutureTask<ReturnType>> {
+ private:
+  std::packaged_task<ReturnType(TaskArgs...)> _task{};  ///< The user-defined C++ task to run
+  std::function<PyObject*(ReturnType)>
+    _pythonConvert{};                 ///< Function to convert the C++ result into Python value
+  PyObject* _asyncioEventLoop{};      ///< The handle to the Python asyncio event loop
+  PyObject* _handle{};                ///< The handle to the Python future
+  std::future<ReturnType> _future{};  ///< The C++ future containing the task result
+
+  /**
+   * @brief Set the result value of the Python future.
+   *
+   * Set the result value of the underlying Python future using the `pythonConvert` function
+   * specified in the constructor to convert the C++ result into the `PyObject*`.
+   *
+   * This function will take the GIL to convert the C++ result into the `PyObject*`.
+   *
+   * @param[in] result the C++ value that will be converted and set as the result of the
+   *                   Python future.
+   */
+  void setResult(const ReturnType result)
+  {
+    // PyLong_FromSize_t requires the GIL
+    if (_handle == nullptr) throw std::runtime_error("Invalid object or already released");
+    PyGILState_STATE state = PyGILState_Ensure();
+    ucxx::python::future_set_result_with_event_loop(
+      _asyncioEventLoop, _handle, _pythonConvert(result));
```

Comment:
Since `future_set_result_with_event_loop()` doesn't have a `noexcept`, we might not release the GIL on failure?

Reply:
There are no C++ exceptions raised in `future_set_result_with_event_loop` (it is supposed to be a pure C function), so I think `noexcept` would be redundant, no?

Reply:
I think `noexcept` would be good both for the sake of clarity and potential compiler optimizations.

Reply:
Just to be sure I'm not misunderstanding your suggestion, https://github.com/rapidsai/ucxx/pull/46/commits/9a7491f2e96d5bee7db70e5d9f125d4841104bce is what you mean, right? You're talking about C++ exceptions and not Python exceptions, the latter might occur if a Python error occurs, e.g. in

Reply:
wrt python exceptions, `_pythonConvert` (which takes the C++ result and produces a Python result) could produce an error (e.g. return `NULL` rather than a valid `PyObject *`) which it looks like we are not handling here?